csrc/megatron/fused_weight_gradient_dense_cuda.cu

```diff
@@ -10,10 +10,481 @@
+        return;
+    }
+
+    uint64_t workspace_size = 0;
+    for(int i = 0; i < returnedAlgoCount; i++)
+        workspace_size = max(workspace_size, heuristicResult[i].workspaceSize);
+    // In this sample, the workspace is already allocated with max_workspace_size
+    // If not, allocate d_workspace here
+    // CHECK_HIP_ERRORhipMalloc(&d_workspace, workspace_size));
+
+    CHECK_HIPBLASLT_ERROR(hipblasLtMatmul(handle,
+                                          matmul,
+                                          &alpha,
+                                          A,
+                                          matA,
+                                          B,
+                                          matB,
+                                          &beta,
+                                          C,
+                                          matC,
+                                          D,
+                                          matD,
+                                          &heuristicResult[0].algo,
+                                          d_workspace,
+                                          workspace_size,
+                                          stream));
+    CHECK_HIPBLASLT_ERROR(hipblasLtMatrixLayoutDestroy(matA));
+    CHECK_HIPBLASLT_ERROR(hipblasLtMatrixLayoutDestroy(matB));
+    CHECK_HIPBLASLT_ERROR(hipblasLtMatrixLayoutDestroy(matC));
+    CHECK_HIPBLASLT_ERROR(hipblasLtMatrixLayoutDestroy(matD));
+    CHECK_HIPBLASLT_ERROR(hipblasLtMatmulDescDestroy(matmul));
+    CHECK_HIPBLASLT_ERROR(hipblasLtMatmulPreferenceDestroy(pref));
+    return;
+}
+
+template <typename T>
+void wgrad_gemm_accum_fp32_cuda(T *input, T *d_output, float *dc_tensor, float *d_weight, int in_dim, int hidden_dim, int out_dim) {
+    //hipblasLtHandle_t handle = at::cuda::getCurrentCUDABlasHandle();
+    //hipStream_t stream;
+    //hipblasGetStream(handle, &stream);
+    hipblasLtHandle_t handle;
+    hipStream_t stream;
+    CHECK_HIP_ERROR(hipStreamCreate(&stream));
+    CHECK_HIPBLASLT_ERROR(hipblasLtCreate(&handle));
+    float alpha = 1.0;
+    float beta  = 1.0;
+    const int batch_count = 1;
+    void*   d_workspace;
+    int64_t max_workspace_size = 32*1024*1024;
+    if(max_workspace_size > 0)
+        hipMalloc(&d_workspace, max_workspace_size);
+    gemmex_wrapper(
+        handle,
+        HIPBLAS_OP_N,
+        HIPBLAS_OP_T,
+        in_dim,      //m
+        out_dim,     //n
+        hidden_dim,  //k
+        batch_count,
+        alpha,
+        beta,
+        input,      //da
+        d_output,   //db
+        dc_tensor,  //dc
+        d_weight,   //dd
+        d_workspace,
+        max_workspace_size,
+        stream);
+
+    CHECK_HIPBLASLT_ERROR(hipblasLtDestroy(handle));
+    CHECK_HIP_ERROR(hipStreamDestroy(stream));
+} 
+    
+template void wgrad_gemm_accum_fp32_cuda<at::Half>(at::Half *input, at::Half *d_output, float *dc_tensor, float *d_weight, int in_dim, int hidden_dim, int out_dim);
```

Comment:
Do we need to add an extra argument of `dc_tensor` here? We can probably use `d_weight` as the input and output.